### ISSUE TITLE
FT_Load_Glyph speedup

### ISSUE THREAD

**Issue by molind** [harfbuzz/harfbuzz#651]
Hello,

We use harfbuzz inside of mobile application Galileo along with SDF renderer. Extremely satisfied by result.

Recently I've looked in profiler to find new bottlenecks and noticed that FT_Load_Glyph causes huge performance slow down on devices with slow internal memory (e.g. iPhone 6). It' takes up to 2 seconds to compute extents and load glyphs information. Mostly because FT_Load_Glyph is slow and calls to FT_Load_Glyph is not cached.

I suggest to use [FreeType cache system](https://www.freetype.org/freetype2/docs/reference/ft2-cache_subsystem.html#FTC_ImageCache_Lookup) instead of reading the same data over and over again. What do you think?

https://github.com/harfbuzz/harfbuzz/blob/20e69c950d431391f92335f8845e5d4291aca8f3/src/hb-ft.cc#L250-L275

https://github.com/harfbuzz/harfbuzz/blob/20e69c950d431391f92335f8845e5d4291aca8f3/src/hb-ft.cc#L294-L322

https://github.com/harfbuzz/harfbuzz/blob/20e69c950d431391f92335f8845e5d4291aca8f3/src/hb-ft.cc#L324-L349

**Comment by khaledhosny** [comment]
Most HarfBuzz clients do some sort of caching so I think you should consider doing the same. Last I tried, just caching `hb_font_t`s resulted in a measurable performance boost.

**Comment by molind** [comment]
@khaledhosny Yes, we do cache `hb_font_t`s. Issue happened when we process some significant amount of text to render e.g. 500 - 5000 words for example. At that moment we have to get glyph extents and other information to draw glyphs correctly.

**Comment by khaledhosny** [comment]
HarfBuzz shouldn’t normally need glyph extents, are you calling `hb_font_get_glyph_extents()` yourself? If so, may be you should cache these.

Anyway, I don’t think we can use Freetype cache API as we don’t create the `FT_Face` ourselves but get it from the client.

**Comment by destman** [comment]
I will try to explain performance problem that we face with more details:
We have 95 fonts that contain all glyphs we need. All this fonts occupy 24MB so instead of loading them into memory we load them from files. In this case FreeType use memory mapping. With new devices all is fine. 

But device with not very fast flash drive have problems with performance.
At our reference test ~4s was spend for text processing. In profiler we find out that our calls of hb_font_get_glyph_extents spend ~2s and shaping spend ~1.9s. 

So we add cache for hb_font_get_glyph_extents (simply std::map<hb_codepoint_t, hb_glyph_extents_t>). This drops time to 18ms. Good result! 

But hb_ft_get_glyph_h_advance still does not cache result of FT_Load_Glyph. 
So we can improve performance even more and question is how to do this better...
Ofc we can set our own function that will return cached horizontal advance (from same cache that we add for hb_font_get_glyph_extents). But maybe it better to do caching at level of harfbuzz (either with std::map or using FreeType caching)...

![screen shot 2017-12-12 at 3 47 18 pm](https://user-images.githubusercontent.com/1651185/33885235-e06fb34e-df53-11e7-9295-cc240ddac9e4.png)

https://github.com/harfbuzz/harfbuzz/blob/20e69c950d431391f92335f8845e5d4291aca8f3/src/hb-ft.cc#L213-L228

**Comment by behdad** [comment]
Right.  This happens if hinting is enabled, as the FT_Get_Advance has to load glyph and apply hints.

I can resurrect `hb_cache_t` and plug it into the advance call.  I had the code and intention to do that but no one ever complained about performance of this code so I removed `hb_cache_t`.  Will give it a try.

**Comment by behdad** [comment]
I'll wait for a Pull Request, or I'll get to this eventually.

**Comment by behdad** [comment]
Closing in preference to the Pull Request.
